### PR TITLE
Docker-build (multi staged)

### PR DESCRIPTION
**Pull request by Napolitain**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Readme.md has been updated to reflect changes.
- [x] Build (`docker build -t noxone/regexgenerator .`) was run locally and any changes were pushed

## Pull request type

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?
Currently, docker is only used to deploy, and you build first with local installation. It means the build doesn't necessarily works on docker, and we have to install gradle to develop.

Issue Number: 1

## What is the new behavior?
We can now build the server image like before by building the project into docker in a multi staged Dockerfile.
It is both safer and more convenient / flexible.

### CODE REVIEW

**Comment by Napolitain** [comment]
Damn, build doesn't work in GA but worked for me. Does that work on your local machine too ?

File: .gitattributes [hunk 1]
```diff
@@ -0,0 +1 @@
+* text=auto eol=lf
```
Comment (Napolitain) [inline .gitattributes:1]: prevent commiting CRLF linebreaks

File: build.gradle [hunk 2]
```diff
@@ -55,7 +55,7 @@ kotlin {
 
             testTask {
                 useKarma {
-                    useChromeHeadless()
+                    useChromeHeadlessNoSandbox()
```
Comment (Napolitain) [inline build.gradle:58]: Required for tests to run in container in root mode

**Comment by noxone** [comment]
Hi, thank you very much for your contributions. Building the page inside a container is a good idea.

> Does that work on your local machine too ?

No, it doesn't. I'm working on an M1 mac and I think that might be the problem. The build using Github Actions creates an AMD64 and an ARM64 image. And it looks like one or more packages you're installing to make the gradle build work, does not exist for ARM64.

I will try to have a look here.

File: .gitattributes (hunk 1 above)
Comment (noxone) [reply .gitattributes:1]: Good idea!

File: build.gradle (hunk 2 above)
Comment (noxone) [reply build.gradle]: I just switched to Firefox, because Chrome seems not to be available on ARM64 platform.

**Comment by noxone** [comment]
OK, now the build works. But I think since it uses QEMU the build process is really slow. 30 Minutes for such a build seems to be too slow for me.

Do you have an idea how to speed this up?

**Comment by Napolitain** [comment]
Did you try this
image
 on macOS ?

**Comment by noxone** [comment]
ah, I don't mean the build on my machine. I mean the build in the Github Actions: [link](https://github.com/noxone/regex-generator/actions/runs/2249733570)

This takes 30 minutes now, which I think is too long. The AMD64 build is still quite fast, but the ARM64 build is horribly slow.

**Comment by noxone** [comment]
Currently the Dockerfile is optimized for use with the Github Actions. It works great when just pushing data into the main branch and then Github Actions will build everything that is needed to publish the web app as well as the docker image.
I understand that if running the Docker build on your local machine, your solution is more generic because the build process itself is kind of independent of the system where it is running.

I would like to provide a really good way to build everything of this project on your local machine. But at the same time I would like to quickly build the web app as well as the published docker images.

And one important thing: I want the whole project to be built when pushing a commit into the repository. And this needs to be fast, in order have a quick feedback whether I broke something. Therefore the Gradle build, as well as the Docker build need to run.

I'm currently thinking of how to achieve this. Maybe we need multiple Docker files building different images and thus use a quicker Docker build process when running in the Github Actions. 
If you have an idea for this, it would be great.

**Comment by Napolitain** [comment]
The newly updated Dockerfile allows use of building directly in Docker using `--build-arg APP_ENV=local` docker parameter and otherwise works as it did before.

**Review by noxone — APPROVED** [review]

**Comment by noxone** [comment]
The latest updates were a really good idea. Thank you for your contributions!

I just googled a little bit and found that we can pass build arguments to the docker build command in the Github Actions, too. So I will try to add the `APP_ENV=github` to the build parameters.
Then the build command in the readme is shorter and it's a bit more convenient for end users.

**Pull request merged**